### ISSUE TITLE
 Using module_trait_correlation to identify gene modules associated with specific disease states.

### ISSUE THREAD

**Issue by Li-mengjie** [smorabit/scWGCNA#242]
Hello, I have been learning about module_trait_correlation (https://smorabit.github.io/hdWGCNA/articles/module_trait_correlation.html), which includes two aspects of significance in the correlation heatmap. It conducts a correlation analysis between gene modules and traits based on cell types. However, I am currently interested in performing a correlation analysis between gene modules and traits based on disease stages, aiming to obtain gene modules specific to certain disease stages. Although I have a general understanding of the analysis process of hdWGCNA, I am still unable to implement my idea mentioned above.

Additionally, I have noticed that when using PlotModuleTraitCorrelation for plotting, it may require traits with 2 or more traits due to subgroup analysis reasons. In reality, I want to obtain a heatmap showing the correlation between multiple disease stages and gene modules, which can include cell types. However, because there is only one trait (cell types) in this scenario, I cannot plot the heatmap. I am not interested in other traits for this analysis.

I look forward to your response.

**Comment by Li-mengjie** [comment]
> Hello, I have been learning about module_trait_correlation (https://smorabit.github.io/hdWGCNA/articles/module_trait_correlation.html), which includes two aspects of significance in the correlation heatmap. It conducts a correlation analysis between gene modules and traits based on cell types. However, I am currently interested in performing a correlation analysis between gene modules and traits based on disease stages, aiming to obtain gene modules specific to certain disease stages. Although I have a general understanding of the analysis process of hdWGCNA, I am still unable to implement my idea mentioned above.
> 
> Additionally, I have noticed that when using PlotModuleTraitCorrelation for plotting, it may require traits with 2 or more traits due to subgroup analysis reasons. In reality, I want to obtain a heatmap showing the correlation between multiple disease stages and gene modules, which can include cell types. However, because there is only one trait (cell types) in this scenario, I cannot plot the heatmap. I am not interested in other traits for this analysis.
> 
> I look forward to your response.

Although I implemented the correlation analysis between disease states and gene modules by setting group.by='orig.ident(disease_state:0_Day,1_Day,30_Day)' parameter, the results contain a lot of NaN and inf values. I'm not sure what went wrong, and I would greatly appreciate it if you could help me with the above issue.
seurat_obj <- ModuleTraitCorrelation(
  seurat_obj,
  traits = cur_traits,
  group.by='orig.ident'
)
![43fa0a445f5ab3ca51f2c70333ec76f](https://github.com/smorabit/hdWGCNA/assets/99384654/28a03daa-f6c0-451d-baec-0e5ac1332fa6)

**Comment by smorabit** [comment]
Could you please share the code that you ran?

**Comment by Li-mengjie** [comment]
> Could you please share the code that you ran?

Of course! Here is the code that I ran for setting the data expression in a Seurat object and grouping cells by 'orig.ident' using SCT assay with metacells enabled. For ease of viewing, I have chosen R script files. Please let me know if there are any other requirements.
[ran code.zip](https://github.com/smorabit/hdWGCNA/files/14644413/ran.code.zip)

**Comment by Li-mengjie** [comment]
> > Could you please share the code that you ran?
> 
> Of course! Here is the code that I ran for setting the data expression in a Seurat object and grouping cells by 'orig.ident' using SCT assay with metacells enabled. For ease of viewing, I have chosen R script files. Please let me know if there are any other requirements. [ran code.zip](https://github.com/smorabit/hdWGCNA/files/14644413/ran.code.zip)

When plotting the correlation heatmap using the PlotModuleTraitCorrelation command, the result is as shown in the figure. Despite grouping by orig.ident (as indicated in the left panel), I still need to set multiple subgroups (traits) to plot correctly（right panel）; otherwise, it will result in the following error. Additionally, as mentioned earlier, the results contain a large number of NaN and inf values.
![9af12cc38fe1a3477c85f764e0d2f40](https://github.com/smorabit/hdWGCNA/assets/99384654/15793b1c-754b-4214-8a87-3df6927692b7)
![293b7ef84150b6301eadcddd75c70dc](https://github.com/smorabit/hdWGCNA/assets/99384654/4f56adcc-4064-402c-b67a-50baa7acf677)
![95d05e731b08303ff0a6ea3c445a79b](https://github.com/smorabit/hdWGCNA/assets/99384654/8938ebaf-053e-4998-b0b7-3350319c59f6)

**Comment by Li-mengjie** [comment]
> > > Could you please share the code that you ran?
> > 
> > 
> > Of course! Here is the code that I ran for setting the data expression in a Seurat object and grouping cells by 'orig.ident' using SCT assay with metacells enabled. For ease of viewing, I have chosen R script files. Please let me know if there are any other requirements. [ran code.zip](https://github.com/smorabit/hdWGCNA/files/14644413/ran.code.zip)
> 
> When plotting the correlation heatmap using the PlotModuleTraitCorrelation command, the result is as shown in the figure. Despite grouping by orig.ident (as indicated in the left panel), I still need to set multiple subgroups (traits) to plot correctly（right panel）; otherwise, it will result in the following error. Additionally, as mentioned earlier, the results contain a large number of NaN and inf values. ![9af12cc38fe1a3477c85f764e0d2f40](https://private-user-images.githubusercontent.com/99384654/313892526-15793b1c-754b-4214-8a87-3df6927692b7.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4OTYyMjIsIm5iZiI6MTcxMDg5NTkyMiwicGF0aCI6Ii85OTM4NDY1NC8zMTM4OTI1MjYtMTU3OTNiMWMtNzU0Yi00MjE0LThhODctM2RmNjkyNzY5MmI3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIwVDAwNTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjM2YzYzljNTEyMzhjM2ViOTkxZmZmYTU3YTNkYzczNjU2ZGRlZTBiYjVjOGU1ODQ1OTJkNjVhMTYyNGJmMzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gU9HLEK1Y0ozB1gNRcavUEAx24t13nbmJqQpPXn_0q8) ![293b7ef84150b6301eadcddd75c70dc](https://private-user-images.githubusercontent.com/99384654/313892584-4f56adcc-4064-402c-b67a-50baa7acf677.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4OTYyMjIsIm5iZiI6MTcxMDg5NTkyMiwicGF0aCI6Ii85OTM4NDY1NC8zMTM4OTI1ODQtNGY1NmFkY2MtNDA2NC00MDJjLWI2N2EtNTBiYWE3YWNmNjc3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIwVDAwNTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2NGJjMTUyZDYxNjViODQ3ZmQ0ZmYyMzRlNGQ2NTQwMzdiNDYyMmE4YzdhMDVkOTAxNzVjZGJmZmU2MzYxM2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PlhUoTXvjB7l68O1NH2ze25qrPdEMAjaXcWNw8BSjSU) ![95d05e731b08303ff0a6ea3c445a79b](https://private-user-images.githubusercontent.com/99384654/313892604-8938ebaf-053e-4998-b0b7-3350319c59f6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4OTYyMjIsIm5iZiI6MTcxMDg5NTkyMiwicGF0aCI6Ii85OTM4NDY1NC8zMTM4OTI2MDQtODkzOGViYWYtMDUzZS00OTk4LWIwYjctMzM1MDMxOWM1OWY2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIwVDAwNTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3Y2ZjZDQwY2U4YzRhYjg1YTNhNmViOWIwOWNiYTMxMDBhYzQ4ODUyN2Q2MDliNjJiMzM0MmRkOWY2NDIyYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XLMNCZ3wst4IdEMaXWw766GmNgaNSsANf88mDB-X31A)

This is my single cell data structure, which I hope will be helpful in understanding my results.
![794a506969e2663fc14fca295396578](https://github.com/smorabit/hdWGCNA/assets/99384654/ce028137-5d9f-432d-9faa-3e3e08776d32)

**Comment by smorabit** [comment]
Thanks for sharing these details. One small request, instead of sharing a zip file can you please include the code directly in this GitHub issue?

I can tell that you have at least one problem. It does not make sense to use `celltype` or `orig.ident` as your traits, since those are categorical variables. Think about traits like age, sex, genotype, disease status. In your case you should use the disease status as one of the traits.

**Comment by Li-mengjie** [comment]
> age, sex, genotype, disease status


Of course, I'm more than willing to share my code directly, but due to its length, I chose to skip the process and only show the code during execution. I agree with your point that using 'celltype' and 'orig.ident' as traits is meaningless. However, as shown in the previous images, my Seurat object does not contain information about age, sex, genotype, etc. The 'disease status' is consistent with 'orig.ident,' so there is no additional column for disease status. Now, what I want to obtain are gene modules related to the disease status (i.e., 'orig.ident'). Do you mean that I need to add an additional column for disease status? In that case, I just need to assign 'orig.ident' to it. However, after trying both methods, I didn't see any difference in results.
<img width="535" alt="image" src="https://github.com/smorabit/hdWGCNA/assets/99384654/00f5637c-5c32-47d9-a469-a5d874add803">
<img width="538" alt="image" src="https://github.com/smorabit/hdWGCNA/assets/99384654/1382bb39-26c5-44ca-bccf-d4fd61ca8ea3">
<img width="571" alt="image" src="https://github.com/smorabit/hdWGCNA/assets/99384654/4bcce0ea-d7a6-4518-b8ef-a7ae00843ff2">

**Comment by smorabit** [comment]
> my Seurat object does not contain information about age, sex, genotype, etc.

To clarify, these are just examples, I you do not have to use those specific traits for this analysis.

> The 'disease status' is consistent with 'orig.ident,' so there is no additional column for disease status. Now, what I want to obtain are gene modules related to the disease status (i.e., 'orig.ident').

When running `ModuleTraitCorrelation`, set `group.by` to something like cluster or cell type, and set `traits` to your traits of interest.

**Comment by Li-mengjie** [comment]
> When running `ModuleTraitCorrelation`, set `group.by` to something like cluster or cell type, and set `traits` to your traits of interest.

As mentioned above, when setting group.by to something like cluster or cell type, and setting traits to your traits of interest, I cannot obtain gene modules related to each time point. Instead, I get gene modules related to the overall time level (as shown in the figure below), which is not what I need. This is because the effects of these time points on cells are bidirectional rather than progressively worsening.
![a914eaf53cbc62ebbca087665670f59](https://github.com/smorabit/hdWGCNA/assets/99384654/693eac23-4733-45bd-ab07-d849b64dea45)

**Comment by smorabit** [comment]
It seems that module trait correlation might not what you are looking for to answer your particular question. I recommend trying [differential module eigengene analysis](https://smorabit.github.io/hdWGCNA/articles/differential_MEs.html#one-versus-all-dme-analysis) instead.

**Comment by nmonikam** [comment]
Hi Sam, I have a question regarding this Module-trait correlation. I have constructed the modules of genes from CD4T only (I extracted the matrix specified for CD4T only). Next, I want to create a heatmap to see the correlation between the CD4T modules with the traits of interest (like age, sex, duration of processing the sample). However, I don't understand why the `ModuleTraitCorrelation` still needs to be grouped by cell types (in `group.by` arg, can I leave it blank?)? How can I make the heatmap that I want (between modules and the traits only)
